### PR TITLE
Offdatabase ordering proposal

### LINKED ISSUE #3608: PG::Error: ERROR: column visualizations.mapviews does not exist
Steps to reproduce:

1. Go to dashboard
1. Order by "visits"

Expected results: List ordered by map views
Actual results: 500 server-side error (see title)

E.g. https://rollbar.com/vizzuality/CartoDB/items/5816/ for more info etc.

### PR DESCRIPTION
**Pull request by juanignaciosl**
@Kartones @rafatower take a look at my proposal for off-database attributes support, which would close #3608 .

The key is implementing in-memory sorting in `Carto:: OffdatabaseQueryAdapter`. That class should implement as many methods as needed, as it mimics an ActiveRecord query. The nice thing of this adapter is that is model-agnostic, you can use it with any class.

I'm not sure whether checking if an attribute is offdatabase or not must be inside the QueryBuilder or the OffdatabaseQueryAdapter.

This is obviously slow as hell for mapviews or likes, since it has to fetch collections or mapviews (in Redis), but optimization can come later.

### CODE REVIEW

**Comment by Cartofante** [comment]
Frontend tests were OK :+1: ([details](http://clinker.cartodb.net/jenkins/job/CartoDB-PR-testing/3649/))

File: app/queries/carto/visualization_query_builder.rb [hunk 1]
```diff
@@ -118,7 +121,12 @@ def with_locked(locked)
   end
 
   def with_order(order, asc_desc = :asc)
-    @order[order] = asc_desc
+    offdatabase_order = offdatabase_order(order)
```
Comment (Kartones) [inline app/queries/carto/visualization_query_builder.rb]: I'd avoid same naming of variables and methods, it kills me :grin:  (just being picky XD)

File: app/queries/carto/visualization_query_builder.rb [hunk 2]
```diff
@@ -209,6 +221,13 @@ def build_paged(page = 1, per_page = 20)
 
   private
 
+  def offdatabase_order(order)
+    return nil unless order.kind_of? String
+    splitted = order.split('.')
+    order_attribute = splitted[splitted.count - 1]
```
Comment (rafatower) [inline app/queries/carto/visualization_query_builder.rb]: just being picky "split" past participle is "split" :p

File: app/queries/carto/visualization_query_builder.rb (hunk 1 above)
Comment (rafatower) [inline app/queries/carto/visualization_query_builder.rb]: offdatabase_order_attribute?

what's the expected format of the parameter `order`

**Comment by rafatower** [comment]
:es:

**Comment by Kartones** [comment]
@javierarce CR please, @javisantana is fleeing from the office right now :disappointed:

**Comment by Kartones** [comment]
Merged into branch `3616-off-db-queries`

File: app/queries/carto/visualization_query_builder.rb (hunk 1 above)
Comment (juanignaciosl) [inline app/queries/carto/visualization_query_builder.rb]: @Kartones you're right, method should be a verb.

File: app/queries/carto/visualization_query_builder.rb (hunk 2 above)
Comment (juanignaciosl) [inline app/queries/carto/visualization_query_builder.rb]: Outch :sweat_smile: . Please change it, @Kartones (I'm assuming you're working in this code).

**Pull request closed without merging**